### DISCUSSION TITLE
Create pdf from html

### DISCUSSION THREAD

**Discussion by webwurm** [AntlerVC/firetable#1109] [Support / Q&A] [answered]
I have a question about how to create a pdf from html. I tried to use the description [here](https://docs.rowy.io/tutorials/pdf-generator) and [here](https://community.flutterflow.io/c/community-tutorials/generate-pdf-part-3-pdf-output-stored-to-firebase-storage) and also checked the example out, but I can't get it to work. Maybe somebody can help.

**The error message is:**
`Failed to launch the browser process! spawn /home/.cache/puppeteer/chrome/linux-1083080/chrome-linux/chrome ENOENT TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md`

**My code:**
```
const derivative:Derivative = async ({row,ref,db,storage,auth,logging})=>{
  logging.log('Task Start: Generating pdf...');

  const puppeteer = require("puppeteer");
  const htmlString = 
  `
  <html>
    <body>
      <h1>{{invoiceNumber}}</h1>
      <p>This is a PDF</p>
    </body>
  </html>
  `;
 
  const replacedHtml = htmlString
    .replace("{{invoiceNumber}}", row.invoiceNumber);

  // take a pdf with populated data 
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.setContent(replacedHtml, { waitUntil: "networkidle0" });
  const pdf = await page.pdf({ printBackground: true });
  await browser.close();

  // upload generated pdf into storage and update output column 
  const pdfFile = await rowy.storage.upload.data(pdf, { 
    fileName: "test.pdf",
    folderPath: `invoicesCreated/${ref.id}`
  });

  return [pdfFile];

}
```

Does anybody have an idea how to make it work? Thanks!

**Comment by BeeBombshell** [comment, marked as answer]
Hey @webwurm, we've experienced similar problems in our demo apps as well. So, as a workaround, you can degrade the puppeteer version like this:
```js
  const puppeteer = require("puppeteer@15.5.0");
```

Check out these demo apps for reference:
- https://rowy.app/p/tryrowy/table/twitterScreenshotBot (extension)
- https://rowy.app/p/tryrowy/table/url2Image
- https://rowy.app/p/tryrowy/table/certificateGenerator

**Reply by webwurm** [reply]
Thanks a lot, but adding this just gives me the error:
`Cannot find module 'puppeteer@15.5.0' Require stack: - /home/node/app/build/scripts/derivative.js - /home/node/app/build/index.js`

I tried the demo apps _(btw.: these links above don't work for me ("You don't have access to the project"), but when I enter them from the Rowy-homepage I can access them)_ - but I think they don't work either. **Url2Image** didn't create screenshots lately - and the last certificate in the **Certificate Generator** was from end of January.

**Reply by BeeBombshell** [reply]
@webwurm, apparently the derivative functions won't work when you'd evaluate. So, you'd have to trigger your functions by the changes on the listener field values. It is generally a good practice to use extensions instead of derivatives for puppeteer apps to avoid such situations.

>Hey @webwurm, we've experienced similar problems in our demo apps as well. So, as a workaround, you can degrade the puppeteer version like this:

Also, I'd like to correct myself, because currently, we don’t support package versioning in the `evaluate` script.

**Reply by webwurm** [reply]
Thank you so much for your help, @BeeBombshell! So I moved it to a Task and now my code looks like this:
```
const extensionBody: TaskBody = async({row, db, change, ref, logging}) => {
  logging.log('Task Start: Generating pdf...');

  const puppeteer = require("puppeteer@15.5.0");
  const htmlString = 
  `
  <html>
    <body>
      <h1>{{invoiceNumber}}</h1>
      <p>This is a PDF</p>
    </body>
  </html>
  `;
 
  const replacedHtml = htmlString
    .replace("{{invoiceNumber}}", row.invoiceNumber);

  // take a pdf with populated data 
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.setContent(replacedHtml, { waitUntil: "networkidle0" });
  const pdf = await page.pdf({ printBackground: true });
  await browser.close();

  // upload generated pdf into storage and update output column 
  try {
    const pdfFile = await rowy.storage.upload.data(pdf, { 
      fileName: "test.pdf",
      folderPath: `invoicesCreated/${ref.id}`
    });

    await ref.update({pdfInvoice: pdfFile});

  } catch (error) {
    logging.error(error);
  }
}
```
Sadly I still don't get a pdf in the pdfInvoice-column and none is generated on the server. I tried to implement some basic error-catching, but in the logs I just see the "Starting task" from the beginning of the task and also no error.

**Reply by webwurm** [reply]
Actually it works for me now - even in the derivative field (although I will run it as a task). I figured out that I didn't set the memory allocation to 1 GB (as it even says in [here](https://docs.rowy.io/tutorials/pdf-generator), but after trying it first then I forgot about it it seems...).

So here is my code that works in the end - in case it helps somebody else:
const derivative:Derivative = async ({row,ref,db,storage,auth,logging})=>{
```
  // Creates PDF
  // only works if memory is increased in the extensions-window
  
  logging.log('Derivative Start: Generating pdf...');

  const puppeteer = require("puppeteer@15.5.0");
  const htmlString = 
  `
  <html>
    <body>
      <h1>{{invoiceNumber}}</h1>
      <p>This is a PDF</p>
    </body>
  </html>
  `;
 
  const replacedHtml = htmlString
    .replace("{{invoiceNumber}}", row.invoiceNumber);

  // take a pdf with populated data 
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.setContent(replacedHtml, { waitUntil: "networkidle0" });
  const pdf = await page.pdf({ printBackground: true });
  await browser.close();

  // upload generated pdf into storage and update output column 
  try {
    const pdfFile = await rowy.storage.upload.data(pdf, { 
      fileName: `${row.invoiceNumber}_${ref.id}.pdf`,
      folderPath: `invoicesCreated`
    });
    logging.log(pdfFile.name);
    return [pdfFile];

  } catch (error) {
    logging.error(error);
  }

}
```

**Reply by BeeBombshell** [reply]
Great catch... Glad it's working now! :)
